src/kernels/jupyter/launcher/serverUriStorage.ts

```diff
@@ -28,16 +31,32 @@ export class JupyterServerUriStorage implements IJupyterServerUriStorage {
     public get onDidRemoveUris() {
         return this._onDidRemoveUris.event;
     }
+    public get currentUri(): string | undefined {
+        return this._currentUri;
+    }
+    public get onDidChange(): Event<void> {
+        return this._onDidChangeUri.event;
+    }
+    public get isLocalLaunch(): boolean {
+        return this.currentUri === Settings.JupyterServerLocalLaunch || this.currentUri === undefined;
```

Comment:
I think we can remove `currentUri` and make a sync version of `isLocalLaunch` available by:
* Returning `true` if we're on web
* Retruning `true` if we have a value in the global memento 
     * I.e. when ever we get/update the encrypted store `await this.encryptedStorage.store(Settings.JupyterServerRemoteLaunchService, key, uri);`, then we know we have a remote, at the same time we can update a memento with a flag.
     * & then use that flag to determine if we have a remote or not & thats a synchronous request, hence we don't need promises & we can still have a non-async version of `isLocalLaunch` without accessing the `this.currentUri`

Reply:
Again, as mentioned earlier, if we access this property too early then we might get th ewrong information, e.g. if `currentUri` hasn't been set we'll get `isLocalLaunch` as true even in web cases (which is incorrect).

Reply:
This race condition was resolved by the `ServerConnectionType` class.

Reply:
And no it wasn't solved by the ServerConnectionType class. It was merely hidden. ServerConnectionType would say not local while the URI wasn't set yet. So if you tried to get the URI, you'd be off of the actual state.

It's why I removed the ServerConnectionType.

Reply:
In fact I think we can get rid of the async version. We don't need it. It could be internal only.

Reply:
> say not local while the URI wasn't set yet. So if you tried to get the URI, you'd be off of the actual state.

We can solve this easily by updating the mememto & a private property.
Hence if the private property has been set, then use that else use memento.
When the class is instantiated the first time (ie. loading VS Code), then property will be `undefined` and we fall back to mememto.

& later on everytime we use the property as that will have the latest state.

Reply:
The whole problem was this state that is mixed between two classes. That seemed broken to me. The true state was in the SErverUriStorage, not the ServerConnectionType.